### PR TITLE
Implement async driver

### PR DESCRIPTION
**Pull request by simonz-bq**

### CODE REVIEW

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (justing-bq) [inline Amazon.QLDB.Driver/driver/QldbDriver.cs:26]: QldbDriver?

File: Amazon.QLDB.Driver/result/AsyncResult.cs
Comment (alpian) [inline Amazon.QLDB.Driver/result/AsyncResult.cs:56]: Explain in docs that this is not thread safe

**Review by alpian — CHANGES_REQUESTED** [review]
The only blocker is the count-then-take unsynchronized block

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriver.cs:26]: "This class or <see cref="AsyncQldbDriver"/> should be..."

I don't see an issue.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:81]: It is needed.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:113]: Done.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:159]: Some encapsulation improvements were made.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:183]: Fixed.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:188]: Updated the message and key a bit to be more inline with what we want

File: Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs:147]: Done.

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriver.cs:284]: Nope 😂

File: Amazon.QLDB.Driver/driver/QldbDriverBase.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriverBase.cs:86]: Updated the comment to be more clear.

File: Amazon.QLDB.Driver/result/AsyncResult.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/result/AsyncResult.cs:56]: Done.

File: Amazon.QLDB.Driver.Tests/AsyncQldbDriverTests.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver.Tests/AsyncQldbDriverTests.cs:58]: While I applaud the cleaning of the code, in the future can we separate the formatting changes from the substantive code changes? It makes it easier to review the actual changes to the code.

File: Amazon.QLDB.Driver/client/Session.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/client/Session.cs:77]: Was this a specific ask to remove the default value? Why?

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/QldbDriver.cs:324]: Can this be replaced with something like:
`return this.driverBase.GetSessionFromPool() ?? this.StartNewSession();`
?

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:198]: What does this mean, we have a transaction issue on starting as session?

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (kylepbit) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:115]: Can we create a ticket to track this so we don't lose the improvement for the future?

File: Amazon.QLDB.Driver.Tests/AsyncQldbDriverTests.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver.Tests/AsyncQldbDriverTests.cs:58]: This is done automatically by settings in my IDE, and when they're not done IntelliSesnse gives you warnings. I think the last person who worked on this did not set this up even though in their environment.

File: Amazon.QLDB.Driver/client/Session.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/client/Session.cs:77]: Putting a default value for the CancellationToken is bad practice when it's on levels where it's being passed through as it allows you to have code that ignores it. The default value should only be on entry-level APIs.

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriver.cs:324]: Done.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:198]: OperationCanceledException is the exception thrown when someone uses a CancellationToken to stop their async operation.

File: Amazon.QLDB.Driver/transaction/Transaction.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/transaction/Transaction.cs:110]: Can we replace the pattern of 
if (x == null) {
 x = not null
}
with
`x ??= not null`

File: Amazon.QLDB.Driver/exception/ExceptionMessages.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/exception/ExceptionMessages.cs:27]: Can we make a ticket to revisit the other drivers to ensure there are no dead messages left?

**Review by kylepbit — COMMENTED** [review]
Overall LGTM - up to you if you do the refactoring in the Execute/retry stuff that we talked about or leave that to the bigger change to state machines.

File: Amazon.QLDB.Driver/exception/ExceptionMessages.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/exception/ExceptionMessages.cs:27]: Done.

File: Amazon.QLDB.Driver/transaction/Transaction.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/transaction/Transaction.cs]: Done

File: Amazon.QLDB.Driver/driver/AsyncQldbDriverBuilder.cs
Comment (alpian) [inline Amazon.QLDB.Driver/driver/AsyncQldbDriverBuilder.cs]: The "Async" part should be a an "extension", not change the name of the user agent.

File: Amazon.QLDB.Driver/driver/QldbDriverBase.cs
Comment (alpian) [inline Amazon.QLDB.Driver/driver/QldbDriverBase.cs]: Rather lock on an internal mutex, not the object itself, which someone else might have locked.

File: Amazon.QLDB.Driver/driver/QldbDriverBase.cs
Comment (alpian) [inline Amazon.QLDB.Driver/driver/QldbDriverBase.cs]: Actually, I think you want `TryTake` instead of Take, then you don't need this lock at all.

File: Amazon.QLDB.Driver/transaction/Transaction.cs
Comment (alpian) [inline Amazon.QLDB.Driver/transaction/Transaction.cs:68]: Same thing, don't lock on something that can be access by the library consumer.

**Review by alpian — CHANGES_REQUESTED** [review]

File: Amazon.QLDB.Driver/driver/AsyncQldbDriverBuilder.cs
Comment (alpian) [inline Amazon.QLDB.Driver/driver/AsyncQldbDriverBuilder.cs:31]: Oh, I see you fixed this

File: Amazon.QLDB.Driver/transaction/AsyncTransaction.cs
Comment (alpian) [inline Amazon.QLDB.Driver/transaction/AsyncTransaction.cs:81]: Rather use `using`.

File: Amazon.QLDB.Driver/driver/QldbDriverBase.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriverBase.cs]: Good suggestion! Done.

File: Amazon.QLDB.Driver/driver/QldbDriverBase.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriverBase.cs]: Lock removed.

File: Amazon.QLDB.Driver/transaction/Transaction.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/transaction/Transaction.cs:68]: Transaction.cs is not accessible by the library consumer.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:136]: Nit - consider changing the name to IsDeadSessionOrThrowIfNoRetryAsync....

Although when put that way, it does seem like two separate things, test if dead, if not dead then check if throwing if no retry....that gets messier and we can probably go down a refactor hole.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:157]: Nice.

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/QldbDriver.cs:293]: I like how much cleaner this became.

File: Amazon.QLDB.Driver/transaction/Transaction.cs
Comment (kylepbit) [reply Amazon.QLDB.Driver/transaction/Transaction.cs:68]: I think the point is that we should lock on an internal, see https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/lock-statement, specifically:

When you synchronize thread access to a shared resource, lock on a dedicated object instance (for example, `private readonly object balanceLock = new object();`) or another instance that is unlikely to be used as a lock object by unrelated parts of the code. Avoid using the same lock object instance for different shared resources, as it might result in deadlock or lock contention. In particular, avoid using the following as lock objects:

- `this`, as it might be used by the callers as a lock.
- Type instances, as those might be obtained by the `typeof` operator or reflection.
- `string` instances, including string literals, as those might be interned.

Hold a lock for as short time as possible to reduce lock contention.

Perhaps lock on this.qldbHash?

File: Amazon.QLDB.Driver/utils/QldbHash.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/utils/QldbHash.cs:70]: What about:
`return ToQldbHash(ValueFactory.NewString(value ?? string.Empty));`

File: Amazon.QLDB.Driver/driver/QldbDriverBase.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/QldbDriverBase.cs:202]: Nit: I think we can probably get rid of the replaceDeadSession local variable here, unless you think it brings readability to the code.

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/QldbDriver.cs:303]: I think GetSession() is only called in one place now, and as a one-liner I wonder if we still need it?

File: Amazon.QLDB.Driver/utils/QldbHash.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/utils/QldbHash.cs:70]: That's a lot better, done.

File: Amazon.QLDB.Driver/driver/QldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriver.cs:303]: It still serves a decent purpose and there's unit tests that depend on it.

File: Amazon.QLDB.Driver/driver/QldbDriverBase.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/QldbDriverBase.cs:202]: Done.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs
Comment (simonz-bq) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:136]: Lets defer this since this will be refactored out later anyways.

**Review by alpian — APPROVED** [review]

**Pull request merged**
